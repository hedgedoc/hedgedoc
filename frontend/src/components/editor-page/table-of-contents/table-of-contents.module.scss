/*
 * SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

.toc {
  width: 100%;
  max-width: 200px;
  max-height: 100vh;
  overflow-y: auto;
  overflow-x: hidden;

  &.sticky {
    position: fixed;
    top: 0;
  }

  padding: 2px;

  > ul > li {
    > a {
      padding: 4px 10px;
    }

    > ul > li {
      > a {
        padding: 1px 0 1px 20px;
      }

      > ul > li {
        > a {
          padding: 1px 0 1px 28px;
        }
      }
    }
  }


  ul {
    padding: 0;
    margin: 0;
    list-style: none;
    font-size: 0.9em;

    a {
      color: #767676;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: pre;
      display: block;
    }


    li.active {
      a {
        margin-top: 2px;
        margin-bottom: 2px;
        color: #000000;
        font-weight: bold;
      }
    }
  }
}
