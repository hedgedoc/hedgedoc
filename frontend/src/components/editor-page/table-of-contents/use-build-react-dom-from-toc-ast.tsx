/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { JumpAnchor } from '../../markdown-renderer/extensions/link-replacer/jump-anchor'
import { tocSlugify } from './toc-slugify'
import type { TocAst } from '@hedgedoc/markdown-it-plugins'
import type { ReactElement } from 'react'
import React, { Fragment, useMemo } from 'react'

/**
 * Generates a React DOM part for the table of contents from the given AST of the document.
 *
 * @param toc The abstract syntax tree of the document for TOC generation
 * @param levelsToShowUnderThis The amount of levels which should be shown below this TOC item
 * @param headerCounts Map that contains the number of occurrences of single header names to allow suffixing them with a number to make them distinguishable
 * @param baseUrl The base URL used for generating absolute links to the note with the correct slug anchor
 */
const buildReactDomFromTocAst = (
  toc: TocAst,
  levelsToShowUnderThis: number,
  headerCounts: Map<string, number>,
  baseUrl: string
): ReactElement | null => {
  if (levelsToShowUnderThis < 0) {
    return null
  }

  const rawName = toc.name.trim()
  const nameCount = (headerCounts.get(rawName) ?? -1) + 1
  const slug = `#${tocSlugify(rawName)}${nameCount > 0 ? `-${nameCount}` : ''}`
  const headlineUrl = new URL(slug, baseUrl).toString()

  headerCounts.set(rawName, nameCount)

  const children = toc.children
    .map((child) => buildReactDomFromTocAst(child, levelsToShowUnderThis - 1, headerCounts, baseUrl))
    .filter((value) => !!value)
    .map((child, index) => <li key={index}>{child}</li>)

  return (
    <Fragment>
      {toc.level > 0 && (
        <JumpAnchor href={headlineUrl} title={rawName} jumpTargetId={slug.slice(1)}>
          {rawName}
        </JumpAnchor>
      )}
      {children.length > 0 && <ul>{children}</ul>}
    </Fragment>
  )
}

/**
 * Generates a React DOM part for the table of contents from the given AST of the document.
 *
 * @param toc The abstract syntax tree of the document for TOC generation
 * @param maxDepth The maximum depth of levels which should be shown in the TOC
 * @param baseUrl The base URL used for generating absolute links to the note with the correct slug anchor
 */
export const useBuildReactDomFromTocAst = (toc: TocAst, maxDepth: number, baseUrl: string) => {
  return useMemo(
    () => buildReactDomFromTocAst(toc, maxDepth, new Map<string, number>(), baseUrl),
    [toc, maxDepth, baseUrl]
  )
}
