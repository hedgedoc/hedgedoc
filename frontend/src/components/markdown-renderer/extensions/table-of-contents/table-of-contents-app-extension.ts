/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { AppExtension } from '../../../../extensions/base/app-extension'
import type { CheatsheetExtension } from '../../../editor-page/cheatsheet/cheatsheet-extension'
import type { MarkdownRendererExtension } from '../base/markdown-renderer-extension'
import { TableOfContentsMarkdownExtension } from './table-of-contents-markdown-extension'
import type EventEmitter2 from 'eventemitter2'

export class TableOfContentsAppExtension extends AppExtension {
  buildMarkdownRendererExtensions(eventEmitter?: EventEmitter2): MarkdownRendererExtension[] {
    return [new TableOfContentsMarkdownExtension(eventEmitter)]
  }

  buildCheatsheetExtensions(): CheatsheetExtension[] {
    return [
      {
        i18nKey: 'toc',
        entries: [
          {
            i18nKey: 'basic'
          },
          {
            i18nKey: 'levelLimit'
          }
        ]
      }
    ]
  }
}
