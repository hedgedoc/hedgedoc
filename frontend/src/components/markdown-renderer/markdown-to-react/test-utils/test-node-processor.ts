/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { NodeProcessor } from '../../node-preprocessors/node-processor'
import type { Document } from 'domhandler'
import { Element, Text } from 'domhandler'

export class TestNodeProcessor extends NodeProcessor {
  process(document: Document): Document {
    document.childNodes.push(new Element('node-processor', {}, [new Text('node processor children')]))
    return document
  }
}
