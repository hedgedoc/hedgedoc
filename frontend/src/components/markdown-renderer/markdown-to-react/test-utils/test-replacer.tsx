/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { NativeRenderer, NodeReplacement, SubNodeTransform } from '../../replace-components/component-replacer'
import { ComponentReplacer, DO_NOT_REPLACE } from '../../replace-components/component-replacer'
import type { Element } from 'domhandler'
import React, { Fragment } from 'react'

export class TestReplacer extends ComponentReplacer {
  replace(node: Element, subNodeTransform: SubNodeTransform, nativeRenderer: NativeRenderer): NodeReplacement {
    return node.tagName === 'node-processor' ? (
      <Fragment>
        <span>NodeProcessor! </span>
        <span data-children={true}> {node.childNodes.map(subNodeTransform)} </span>
        <span data-native={true}> {nativeRenderer()} </span>
      </Fragment>
    ) : (
      DO_NOT_REPLACE
    )
  }
}
