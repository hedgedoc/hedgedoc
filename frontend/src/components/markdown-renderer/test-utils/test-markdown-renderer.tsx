/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { StoreProvider } from '../../../redux/store-provider'
import type { MarkdownRendererExtension } from '../extensions/_base-classes/markdown-renderer-extension'
import { MarkdownToReact } from '../markdown-to-react/markdown-to-react'
import React, { useMemo } from 'react'

export interface SimpleMarkdownRendererProps {
  content: string
  extensions: MarkdownRendererExtension[]
}

/**
 * A markdown renderer for tests.
 *
 * @param content The content to be rendered.
 * @param extensions The {@link MarkdownRendererExtension MarkdownExtensions} to use for rendering.
 */
export const TestMarkdownRenderer: React.FC<SimpleMarkdownRendererProps> = ({ content, extensions }) => {
  const lines = useMemo(() => content.split('\n'), [content])
  return (
    <StoreProvider>
      <MarkdownToReact
        markdownContentLines={lines}
        markdownRenderExtensions={extensions}
        newlinesAreBreaks={true}
        allowHtml={false}
      />
    </StoreProvider>
  )
}
