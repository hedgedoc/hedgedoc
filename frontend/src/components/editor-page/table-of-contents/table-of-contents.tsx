/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { concatCssClasses } from '../../../utils/concat-css-classes'
import { ShowIf } from '../../common/show-if/show-if'
import styles from './table-of-contents.module.scss'
import { useBuildReactDomFromTocAst } from './use-build-react-dom-from-toc-ast'
import type { TocAst } from '@hedgedoc/markdown-it-plugins'
import React from 'react'
import { Trans, useTranslation } from 'react-i18next'

export interface TableOfContentsProps {
  ast: TocAst
  maxDepth?: number
  className?: string
  baseUrl: string
}

/**
 * Renders the table of contents for the note.
 *
 * @param ast The ast of the headings in the current note.
 * @param maxDepth The maximal depth of that is considered in the tree
 * @param className Additional classes directly given to the div
 * @param baseUrl The base url of the instance.
 */
export const TableOfContents: React.FC<TableOfContentsProps> = ({ ast, maxDepth = 3, className, baseUrl }) => {
  useTranslation()
  const tocTree = useBuildReactDomFromTocAst(ast, maxDepth, baseUrl)

  return (
    <div className={concatCssClasses(styles.toc, className)}>
      <ShowIf condition={ast.children.length === 0}>
        <Trans i18nKey={'editor.infoToc'} />
      </ShowIf>
      {tocTree}
    </div>
  )
}
