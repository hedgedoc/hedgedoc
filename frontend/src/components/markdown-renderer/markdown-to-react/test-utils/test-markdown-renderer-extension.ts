/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { MarkdownRendererExtension } from '../../extensions/_base-classes/markdown-renderer-extension'
import type { NodeProcessor } from '../../node-preprocessors/node-processor'
import type { ComponentReplacer } from '../../replace-components/component-replacer'
import { TestNodeProcessor } from './test-node-processor'
import { TestReplacer } from './test-replacer'
import type MarkdownIt from 'markdown-it'
import Token from 'markdown-it/lib/token'

export class TestMarkdownRendererExtension extends MarkdownRendererExtension {
  buildNodeProcessors(): NodeProcessor[] {
    return [new TestNodeProcessor()]
  }

  configureMarkdownIt(markdownIt: MarkdownIt) {
    markdownIt.use(() => {
      markdownIt.core.ruler.push('configure', (core) => core.tokens.push(new Token('configure', 'configure', 0)))
      markdownIt.renderer.rules.configure = () => '<span>configure</span>'
    })
  }

  buildReplacers(): ComponentReplacer[] {
    return [new TestReplacer()]
  }

  buildTagNameAllowList(): string[] {
    return ['node-processor']
  }

  configureMarkdownItPost(markdownIt: MarkdownIt) {
    markdownIt.use(() => {
      markdownIt.core.ruler.push('post', (core) => core.tokens.push(new Token('post', 'post', 0)))
      markdownIt.renderer.rules.post = () => '<span>post</span>'
    })
  }
}
