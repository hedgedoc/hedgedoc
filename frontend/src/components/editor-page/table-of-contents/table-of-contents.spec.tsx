/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { mockI18n } from '../../../test-utils/mock-i18n'
import { TableOfContents } from './table-of-contents'
import type { TocAst } from '@hedgedoc/markdown-it-plugins'
import { render } from '@testing-library/react'

describe('Table of contents', () => {
  beforeAll(async () => {
    await mockI18n()
  })

  const level4Ast: TocAst = {
    name: 'Level 4',
    level: 4,
    children: []
  }
  const level3Ast: TocAst = {
    name: 'Level 3',
    level: 3,
    children: [level4Ast]
  }
  const level2Ast: TocAst = {
    name: 'Level 2',
    level: 2,
    children: [level3Ast]
  }
  const level1Ast: TocAst = {
    name: 'Level 1',
    level: 1,
    children: [level2Ast]
  }
  const level0Ast: TocAst = {
    name: '',
    level: 0,
    children: [level1Ast]
  }

  it('renders correctly', () => {
    const view = render(
      <TableOfContents ast={level0Ast} className={'customClassName'} baseUrl={'https://example.org'} />
    )
    expect(view.container).toMatchSnapshot()
  })

  it('renders only in requested max depth', () => {
    const view = render(
      <TableOfContents ast={level0Ast} maxDepth={2} className={'customClassName'} baseUrl={'https://example.org'} />
    )
    expect(view.container).toMatchSnapshot()
  })
})
