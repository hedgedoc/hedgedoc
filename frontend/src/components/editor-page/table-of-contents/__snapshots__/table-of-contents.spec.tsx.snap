// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table of contents renders correctly 1`] = `
<div>
  <div
    class="markdown-toc customClassName"
  >
    <ul>
      <li>
        <a
          href="https://example.org/#level-1"
          title="Level 1"
        >
          Level 1
        </a>
        <ul>
          <li>
            <a
              href="https://example.org/#level-2"
              title="Level 2"
            >
              Level 2
            </a>
            <ul>
              <li>
                <a
                  href="https://example.org/#level-3"
                  title="Level 3"
                >
                  Level 3
                </a>
              </li>
            </ul>
          </li>
        </ul>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Table of contents renders only in requested max depth 1`] = `
<div>
  <div
    class="markdown-toc customClassName"
  >
    <ul>
      <li>
        <a
          href="https://example.org/#level-1"
          title="Level 1"
        >
          Level 1
        </a>
        <ul>
          <li>
            <a
              href="https://example.org/#level-2"
              title="Level 2"
            >
              Level 2
            </a>
          </li>
        </ul>
      </li>
    </ul>
  </div>
</div>
`;
